Add CSV as output format

sedrubal opened the pull request.

criztovyl commented on firefox_decrypt.py:
```diff
@@ -751,8 +761,9 @@ def parse_sys_args():
     parser.add_argument("-p", "--pass-prefix", action="store", default=u"web",
                         help="Prefix for export to pass from passwordstore.org (default: %(default)s)")
-    parser.add_argument("-t", "--tabular", action="store_true",
-                        help="Output in tabular format")
```
I would either keep this (in a backwards-compatible way) or fix the corresponding test (verifying it still does it's job. ^^).
Also: Maybe add a test for CSV?

sedrubal replied:
I fixed the tests in 3488d77 but I'll try to keep backwards compatibility like @unode suggests.

sedrubal replied:
Done with dd112d6. New tests are missing...

criztovyl commented on firefox_decrypt.py:
```diff
@@ -430,6 +431,14 @@ def output_line(line):
+            csv_writer = csv.DictWriter(
+                sys.stdout, fieldnames={'host', 'user', 'password'},
+                delimiter=';', quotechar='|', quoting=csv.QUOTE_ALL
+            )
```
I think `"` is a more suitable quote-char. Maybe expose this to the user somehow?

sedrubal replied:
I'll add a `-q` / `--quotechar` argument.

unode commented on firefox_decrypt.py:
```diff
@@ -430,6 +431,14 @@ def output_line(line):
+        if format == 'csv':
+            csv_writer = csv.DictWriter(
+                sys.stdout, fieldnames={'host', 'user', 'password'},
+                delimiter=';', quotechar='|', quoting=csv.QUOTE_ALL
```
I have a bit of a problem with this line. CSV is not a standard format and using `;` together with `|` is definitely not standard.

sedrubal replied:
Yes you're right. I'll change this.

sedrubal commented on firefox_decrypt.py:
```diff
@@ -409,7 +410,7 @@ def decode_entry(self, user64, passw64):
         return user, passw
 
-    def decrypt_passwords(self, export, tabular):
+    def decrypt_passwords(self, export, format='human'):
```
Yes I'm sorry.

sedrubal commented on firefox_decrypt.py:
```diff
@@ -430,6 +431,14 @@ def output_line(line):
+                sys.stdout, fieldnames={'host', 'user', 'password'},
+                delimiter=';', quotechar='|', quoting=csv.QUOTE_ALL
+            )
+            if header:
```
I think this bug fix will be deprecated with dd112d6. Am I right?

sedrubal commented:
I think I've solved all thins from you review (thanks by the way :wink: ). But we have to change the tests again and add new test...

unode commented:
@sedrubal the tests should stay the same to ensure backwards compatibility but I agree that new tests should be added to ensure new functionality works as expected.

Do you think you can write these tests yourself? It should be similar to the existing tabular test which is currently still failing.

unode commented:
(Sorry the last comments should have been a single review)

unode commented on CHANGELOG.md:
```diff
@@ -3,6 +3,7 @@
 ##### git
 - Fix PK11 slot memory leak
 - Configurable pass-export prefix via `--pass-prefix`
+- Remove `--tabular` and add `--format` parameter and support also CSV format
```
`Deprecate` instead of `Remove`, we'll keep it for a while.

sedrubal replied:
👍

unode commented on firefox_decrypt.py:
```diff
@@ -768,6 +779,10 @@ def parse_sys_args():
 
+    if args.tabular:
+        args.format = 'csv'
+        args.delimiter = '\t'
```
Minor, `"` here too instead of `'`.

unode replied:
Missing args.quotechar = "'" to ensure backwards compatibility.

I'm actually not entirely sure what the csv module does if there's a ' symbol in a password. It might escape it somewhow. A test should also cover this case.

sedrubal replied:
Yes it does escape them (that's why I implemented this at all :wink:). I tried to add a test case for this but I didn't know how to edit the firefox password files.

unode replied:
Yes, but for backwards compatibility we should keep `'` as the quotechar when outputting the `tabular` format. The old tabular used `\t` and `'` instead of `\t` and `"`.

The other comment was related to using `"` for strings instead of `'` to be consistent with the rest of the script.

sedrubal replied:
Should be fixed with 97ff93f

unode commented on tests/bash_tap_fd.sh:
```diff
@@ -10,6 +10,10 @@ function get_script {
     echo "/usr/bin/env python ${bashtap_org_pwd}/../firefox_decrypt.py"
 }
 
+function get_script_python3 {
```
This isn't necessary, use `get_script_python` instead.
We run tests on different machines. One has python 2 as default others have python 3 so `python` can either be 2 or 3 depending on where tests run.

sedrubal replied:
Should be fixed with a075320

unode commented on tests/tabular_46.t:
```diff
@@ -2,7 +2,7 @@
 
 PASSWD=$(get_password)
-CMD="$(get_script) --tabular"
+CMD="$(get_script) --format tabular"
```
Keep a copy of the `--tabular` test for backwards compat. testing. Call it `tabular_deprecated_46.t`
The new test looks good.

unode commented on tests/csv_46.t:
```diff
@@ -0,0 +1,12 @@
+CMD="$(get_script_python3) --format csv"
+TEST="$(get_test_data)/test_profile_firefox_46/"
+
+${bashtap_org_pwd}/diff_csv.py <(echo ${PASSWD} | ${CMD} ${TEST} | grep doesntexist || kill $$) <(get_user_data "doesntexist_csv")
```
Feel free to use the same 'dumb' approach from before of using `diff -u`. No need to check if the output is parseable since other users may not be using something else to read the csv format.
Bruteforcing on the output being exactly what is in the expected file should be sufficient.

unode replied:
A better test would be to test different `--delimiter` and `--quotechar` from default settings.
This ensures those two arguments are working as expected.

unode commented on firefox_decrypt.py:
```diff
@@ -796,6 +811,8 @@ def main():
 
     setup_logging(args)
+    if args.tabular:
+        LOG.warning("--tabular is deprecated. Use `--format csv --delimiter \\t` instead")
```
This looks good. Thanks!

unode commented on firefox_decrypt.py:
```diff
@@ -409,7 +410,7 @@ def decode_entry(self, user64, passw64):
         return user, passw
 
-    def decrypt_passwords(self, export, tabular):
+    def decrypt_passwords(self, export, format="human", csv_delimiter=";", csv_quotechar="|"):
```
Perhaps `output_format` instead of `format`. `format` is actually a function on python 3.

sedrubal replied:
👍

unode commented on firefox_decrypt.py:
```diff
@@ -746,14 +752,19 @@ def parse_sys_args():
+                        default="human", help="Format for the output.")
+    parser.add_argument("-d", "--delimiter", action="store", default=";",
+                        help="The delimiter for csv output")
+    parser.add_argument("-q", "--quotechar", action="store", default="|",
```
Can you set this default to `"` instead of `|` to use slightly more common CSV settings?

sedrubal replied:
ok.

unode replied:
You can use `'"'` in this case to avoid having to do `"\""` which is much less readable.

unode commented on firefox_decrypt.py:
```diff
@@ -816,7 +833,9 @@ def main():
     # Decode all passwords
-    to_export = nss.decrypt_passwords(args.export_pass, args.tabular)
+    to_export = nss.decrypt_passwords(
+        args.export_pass, args.format, csv_delimiter=args.delimiter, csv_quotechar=args.quotechar,
```
This is incorrect and shouldn't be necessary. Exporting to pass is unrelated to printing to stdout.

sedrubal replied:
What exactly do you mean? I think it's the same (dirty) way as it was used before for `tabular`.

unode replied:
Sorry, please ignore this comment. I read it as referring to a different part of the code.

unode commented:
Looking good so far. Thanks for the effort.

sedrubal commented:
Currently I've to learn for my exams, but I'll complete this pr this soon ™️

sedrubal commented:
I think I fixed all things you requested in the review but there are still 2 "bugs":

- The first line in the output of `--tabular` changed:
```diff
- Website	Username	Password
+ 'host'	'user'		'password'
```
This wasn't covered by the test cases so I assume this is ok :smile:

- As I didn't manage to solve utf-8 incompatibility when using CSV, all test cases using python2 currently fail. To be honest, I don't want to fix this for python2 because IMHO there is no use in supporting python2 anymore...

unode commented on tests/test_data/users/complex_csv_a_b.user:
```diff
@@ -0,0 +1 @@
+bhttps://githubb.combabcömplexbabсЮЛОажс$4vz*VçàhxpfCbbmwob
```
This one made me laugh :smile: but I think we don't need it.
Better cover the `,` and `'` use-case instead.
If you remove this it's also fine since the rest of the tests are quite extensive already.

sedrubal replied:
I did this, because no password did contain any of the usual delimiter or quote characters. But yes, we can remove this now 😆 thanks for pushing one...

unode commented on tests/csv_46.t:
```diff
@@ -0,0 +1,41 @@
+CMD="$(get_script) --format csv"
+TEST="$(get_test_data)/test_profile_firefox_46/"
+
+diff -u <(echo ${PASSWD} | ${CMD} ${TEST} | head -n 1 || kill $$) <(get_user_data "header_csv_default")
```
This is nice. The header wasn't tested before.

sedrubal commented on firefox_decrypt.py:
```diff
@@ -430,6 +430,16 @@ def output_line(line):
+            if not PY3:
+                LOG.warning("csv format is only supported with python3. The script may crash.")
+            csv_writer = csv.DictWriter(
+                sys.stdout, fieldnames=["host", "user", "password"],
```
Fixed with 9ff8bac.

sedrubal commented on firefox_decrypt.py:
```diff
@@ -751,8 +761,9 @@ def parse_sys_args():
                         help="Export URL, username and password to pass from passwordstore.org")
     parser.add_argument("-p", "--pass-prefix", action="store", default=u"web",
                         help="Prefix for export to pass from passwordstore.org (default: %(default)s)")
-    parser.add_argument("-t", "--tabular", action="store_true",
```
53e5b4b

sedrubal commented:
Ok, I rebased and tests are working for python3... (I also tried to adapt the commit messages to your scheme).

Another possible problem is, that the csv module uses 'CRLF' instead of 'LF'. `--tabular` used `LF`. I hope this is ok... (I changed all tabular test cases to dos format).

unode commented:
> Another possible problem is, that the csv module uses CRLF instead of LF. --tabular used LF. I hope this is ok... (I changed all tabular test cases to dos format).

https://stackoverflow.com/a/9845731 might fix that.

For the python2 utf8 issue. I'll have to try and see.
There is a third-party unicodecsv module but I'd prefer to not add any dependency outside stdlib.

unode commented:
Finally, feel free to squash most of the commits.
No need to track the intermediate steps during review/discussion.

I'll play with the csv part a bit before merging.

sedrubal commented:
Usually I like to keep the history, but in this case I can do it. The fix was too easy, thanks 😄

unode commented:
Or not, the fix didn't work for py3. *sigh*
We'll need an if PY3 around that.

sedrubal commented:
Ok, I'll do it while sqashing...

unode commented:
> https://stackoverflow.com/a/9845731 might fix that

Can you also do this one.

We are all green so good to go.

unode commented:
Merged manually fc66e08. Great work.

Thanks again for all the effort and bearing with me on all the minor details.

sedrubal commented:
Thanks. It's ok. A good project needs strict reviews :smile:

Should I still set the `lineterminator` to `\n`? I think the python developers will know what they're doing, so there might be reasons for `\r\n`...

unode commented:
I did the lineterminator change as part of the manual merge.

You have a point on reasons for `\r\n` but I've never encountered such case while working on Linux. Even Open/LibreOffice defaults to `\n` so I prefer to be consistent in all output.  
We already have `\n` on stdout.

If I ever get to automate tests on a Windows platform I'll reconsider.
